File: android/src/main/kotlin/com/tqc/ads/flutter_admob_native_ads/banner/BannerAdPlatformView.kt
package com.tqc.ads.flutter_admob_native_ads.banner

import android.content.Context
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.google.android.gms.ads.AdView
import com.tqc.ads.flutter_admob_native_ads.FlutterAdmobNativeAdsPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

/**
 * Platform view for displaying banner ads.
 *
 * This view uses the shared banner ad from the plugin's BannerAdLoader
 * instead of creating and loading its own ad. This prevents duplicate
 * ad requests and ensures the ad is loaded only once per controller.
 */
class BannerAdPlatformView(
    private val context: Context,
    private val viewId: Int,
    private val creationParams: Map<String, Any?>,
    private val messenger: BinaryMessenger
) : PlatformView {

    companion object {
        private const val TAG = "BannerAdPlatformView"
        private const val BANNER_CHANNEL_NAME = "flutter_admob_banner_ads"
    }

    private val container: FrameLayout = FrameLayout(context).apply {
        layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        )
    }

    private val enableDebugLogs: Boolean
    private val controllerId: String?
    private val channel: MethodChannel

    init {
        enableDebugLogs = creationParams["enableDebugLogs"] as? Boolean ?: false
        controllerId = creationParams["controllerId"] as? String
        channel = MethodChannel(messenger, BANNER_CHANNEL_NAME)

        log("Initializing banner platform view for controller: $controllerId")

        // Register to receive the shared banner ad from the loader
        registerForBannerAd()
    }

    /**
     * Registers a callback to receive the banner ad from the shared loader.
     * Also checks if an ad is already loaded and uses it immediately.
     */
    private fun registerForBannerAd() {
        val plugin = FlutterAdmobNativeAdsPlugin.getInstance()

        if (controllerId == null) {
            log("Invalid controllerId")
            return
        }

        log("Registering for banner ad with controller: $controllerId")

        // Register callback to receive ad when it loads
        plugin?.registerBannerAdCallback(controllerId!!) { bannerView ->
            onBannerLoaded(bannerView)
        }

        // Check if ad is already loaded
        val existingBanner = plugin?.getBannerAd(controllerId!!)
        if (existingBanner != null) {
            log("Banner already loaded, using existing ad view")
            onBannerLoaded(existingBanner)
        } else {
            log("No banner loaded yet, waiting for callback")
        }
    }

    /**
     * Called when the banner ad is loaded.
     * Adds the ad view to the container.
     */
    private fun onBannerLoaded(bannerView: AdView) {
        log("Banner received, adding to container")

        // Remove from previous parent if any
        (bannerView.parent as? ViewGroup)?.removeView(bannerView)

        // Clear container and add the banner view
        container.removeAllViews()
        container.addView(bannerView, FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        ))

        log("Banner view added to container")
    }

    /**
     * Sends an event to Flutter via method channel.
     */
    private fun sendEvent(method: String, arguments: Map<String, Any?>) {
        try {
            channel.invokeMethod(method, arguments)
        } catch (e: Exception) {
            log("Error sending event $method: ${e.message}")
        }
    }

    override fun getView(): View = container

    override fun dispose() {
        log("Disposing banner platform view")

        // Unregister callback
        controllerId?.let {
            FlutterAdmobNativeAdsPlugin.getInstance()?.unregisterBannerAdCallback(it)
        }

        // Remove view from container but don't destroy the ad
        // (it's managed by the BannerAdLoader)
        container.removeAllViews()
    }

    private fun log(message: String) {
        if (enableDebugLogs) {
            Log.d(TAG, "[$viewId] $message")
        }
    }
}

File: ios/Classes/BannerAd/BannerAdPlatformView.swift
import Flutter
import UIKit
import GoogleMobileAds

/// Platform view for displaying banner ads.
class BannerAdPlatformView: NSObject, FlutterPlatformView {

    private let containerView: UIView
    private var bannerView: GADBannerView?
    private let enableDebugLogs: Bool
    private let controllerId: String?

    init(
        frame: CGRect,
        viewId: Int64,
        creationParams: [String: Any],
        messenger: FlutterBinaryMessenger
    ) {
        self.containerView = UIView(frame: frame)
        self.enableDebugLogs = creationParams["enableDebugLogs"] as? Bool ?? false
        self.controllerId = creationParams["controllerId"] as? String

        super.init()

        containerView.backgroundColor = .clear

        log("Initializing banner platform view")

        registerForAdUpdates()
    }

    func view() -> UIView {
        return containerView
    }

    /// Registers with the plugin to receive ad updates.
    /// IMPORTANT: Register callback FIRST, then check existing ad to avoid race condition.
    private func registerForAdUpdates() {
        guard let controllerId = controllerId else {
            log("Invalid controllerId, cannot register for ad updates")
            return
        }

        // Register callback FIRST to avoid missing ads that load between check and register
        FlutterAdmobNativeAdsPlugin.shared()?.registerBannerAdCallback(controllerId: controllerId) { [weak self] bannerView in
            self?.onAdLoaded(bannerView)
        }

        // THEN check if ad is already loaded (from cache)
        if let existingBannerView = FlutterAdmobNativeAdsPlugin.shared()?.getBannerAd(controllerId: controllerId) {
            onAdLoaded(existingBannerView)
        }
    }

    private func onAdLoaded(_ view: GADBannerView) {
        // Remove old banner view if exists
        if let oldBanner = bannerView {
            oldBanner.removeFromSuperview()
        }

        bannerView = view

        containerView.addSubview(view)

        view.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            view.topAnchor.constraint(equalTo: containerView.topAnchor),
            view.leadingAnchor.constraint(equalTo: containerView.leadingAnchor),
            view.trailingAnchor.constraint(equalTo: containerView.trailingAnchor),
            view.bottomAnchor.constraint(equalTo: containerView.bottomAnchor)
        ])
    }

    private func log(_ message: String) {
        if enableDebugLogs {
            print("[BannerAdPlatformView] \(message)")
        }
    }

    deinit {
        if let controllerId = controllerId {
            FlutterAdmobNativeAdsPlugin.shared()?.unregisterBannerAdCallback(controllerId: controllerId)
        }
    }
}
